### PR TITLE
Whitelist Az module for culture directory skipping

### LINKED ISSUE #8125: Get-Module -ListAvailable skips module found in PSModulePath
Installing the `Az` module from the PowerShell Gallery will install the base `Az` module, as well as all other submodules (_e.g._, `Az.Compute`, `Az.Sql`, etc.) that it requires -- however, this module appears to be skipped when running `Get-Module -ListAvailable`, whereas all of the submodules appear.

Steps to reproduce
------------------

```powershell
> Install-Module -Name Az -Repository PSGallery -Force

> Get-Module -Name Az -ListAvailable
```

Expected behavior
-----------------

```none
> Get-Module -Name Az -ListAvailable


    Directory: C:\Program Files\PowerShell\Modules


ModuleType Version    Name                                PSEdition ExportedCommands
---------- -------    ----                                --------- ----------------
Script     0.4.0      Az                                  Core,Desk 

>
```

Actual behavior
---------------

```none
> Get-Module -Name Az -ListAvailable
>
```

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.0
PSEdition                      Core
GitCommitId                    6.1.0
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Fixes https://github.com/PowerShell/PowerShell/issues/8125.

Takes out the optimisation to skip directories named like cultures when listing modules. If we have a better way to preserve some of the performance benefits, I can happily implement!

Also generalised some of the test code a bit. Realised afterward that it wasn't as helpful as I thought, but it's done now at least. Commits are separate, so that can be dumped if we want.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

I haven't marked "Not Breaking" yet, since technically this is a change in outward behaviour. But I don't think this behaviour is documented and I doubt anyone knew about it let alone relied upon it. But want to leave that judgement for others to decide.

### CODE REVIEW

**Comment by anmenaga** [comment]
Restarted `PowerShell-CI-linux` & `PowerShell-CI-macos` test runs.

**Comment by SteveL-MSFT** [comment]
@rjmholt can you do a perf test on Windows to see if there's any noticeable degradation?

**Comment by lzybkr** [comment]
IIRC, this was pretty important for performance. The culture directories sometimes have a bunch of `psd1` files which we analyze as a module manifest and bail when we realize they aren't.

Some options I can think of
1. Check the folder depth first, skipping the culture check only for folders immediately below a path in PSModulePath.
1. Process culture folders, but cache the "failure" somehow - messy and not helpful for first ever start, e.g. in a container
1. Whitelist `az`
1. Rename the `az` module

I'm curious, does the `az` module have localized resources for the `az` culture? It'd be good to test that works as well.

**Comment by vexx32** [comment]
Given the commonality of the `Az` module I wouldn't be surprised if someone else creates a module with a similarly-styled name, if one or more doesn't already exists. Folks tend to copy anything with any kind of an 'official' stamp on it, after all.

Whitelisting may be the simplest option for now, but unless we want to *forbid* future modules with such short names / folder names, is there a more elegant and generalisable solution available?

**Comment by rjmholt** [comment]
Noted about the perf.

Two lines of questioning so I can try and understand the general problem better:
- Why does this "lots of culture dirs on the module path" phenomenon occur? Is it a specific other scenario/application that's using them? Why do they contain `psd1` files (and are they not PowerShell manifests)? Is any of this written up or coded anywhere?
- Is there an example anyone has of such a directory that has both modules and culture dirs I could look at?

**Comment by lzybkr** [comment]
Does this just affect autoloading?

And what servicing 5.1?

**Comment by SteveL-MSFT** [comment]
@lzybkr I don't believe this qualifies to service 5.1.  The Az module is still loaded and works, this only affects listing the loaded modules.

**Comment by rjmholt** [comment]
@lzybkr The issue seems to be with `Get-Module -ListAvailable` (see https://github.com/Azure/azure-powershell/issues/7563), although I wouldn't be surprised if that meant there are issues elsewhere.

**Comment by lzybkr** [comment]
It is unlikely I would have added that code to make `Get-Module -ListAvailable` faster, I was focused on first ever PowerShell startup when that code was written.

I'm pretty sure that code is called during module analysis for module auto-loading.

**Comment by rjmholt** [comment]
Ok, I'm personally leaning toward whitelisting just `Az` at the moment, although I'm not a huge fan of hardcoding it.

**Comment by rjmholt** [comment]
I've implemented a whitelist. I've left a note on the whitelist describing a workaround for other modules in future if needed.

**Comment by adityapatwardhan** [comment]
@rjmholt  Please have a look at CodeFactor issue.

**Comment by adityapatwardhan** [comment]
Restarted CI for pick up build failure test fix.

File: src/System.Management.Automation/engine/Modules/ModuleUtils.cs [hunk 1]
```diff
@@ -24,6 +24,16 @@ internal static class ModuleUtils
...
                                         new System.IO.EnumerationOptions() { AttributesToSkip = 0, BufferSize = 16384 };
 
+        /// <summary>
+        /// Whitelist for modules whose names collide with culture names, so that they aren't skipped
+        /// by IsPossibleModuleDirectory(). While this is private, having a static HashSet means that
+        /// further modules could be added using reflection in case a workaround is needed.
+        /// </summary>
+        private static readonly HashSet<string> ModuleNameCultureWhitelist = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "Az"
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/Modules/ModuleUtils.cs]: Perhaps add a comment that this is for the Azure PowerShell module

File: src/System.Management.Automation/engine/Modules/ModuleUtils.cs [hunk 2]
```diff
@@ -24,6 +24,16 @@ internal static class ModuleUtils
         private readonly static System.IO.EnumerationOptions s_uncPathEnumerationOptions =
                                         new System.IO.EnumerationOptions() { AttributesToSkip = 0, BufferSize = 16384 };
 
+        /// <summary>
+        /// Whitelist for modules whose names collide with culture names, so that they aren't skipped
+        /// by IsPossibleModuleDirectory(). While this is private, having a static HashSet means that
+        /// further modules could be added using reflection in case a workaround is needed.
+        /// </summary>
+        private static readonly HashSet<string> ModuleNameCultureWhitelist = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/Modules/ModuleUtils.cs]: I think for private members you should still use the underscore prefix

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by lzybkr** [comment]
The `s_` prefix was introduced when running the Roslyn code formatter and it seems to still be used heavily, so I'm not sure why Code Factor wins out on this one.

**Comment by rjmholt** [comment]
I also like (*prefer*) `s_`. I find it very informative, we use it everywhere, and it meshes nicely with `_` for non-static Private.

**Comment by vexx32** [comment]
Maybe we ought to add a StyleCop exemption for this sort of thing? I've run into it before as well.

**Comment by SteveL-MSFT** [comment]
I believe CodeFactor is using the [StyleCop](https://github.com/PowerShell/PowerShell/blob/53e6ec6eadc2f363fc671e96b34ade4dda3fe8eb/Settings.StyleCop) rules we've set.  I'm ok with changing the rules to allow for `s_`.  My main desire is to have consistency so if most of the code is using `s_` we should adopt that.

**Comment by iSazonov** [comment]
@rjmholt Please use `s_` and we'll look which rules we should fix.

**Comment by rjmholt** [comment]
Thanks @iSazonov!

**Comment by anmenaga** [comment]
Restarted appveyor test pass for WebListener failure.

**Comment by rjmholt** [comment]
Only remaining checks are the CodeFactor rules we agreed to ignore.

**Comment by iSazonov** [comment]
Because of we do `new CultureInfo(dir)`  my suggestion is to use on-demand cache for GetCultureInfo() (we could reuse it in Get-Culture cmdlet and vise versa) and then build a HashSet for culture names.  

@mklement0 Could you please look the PR? Again cultureinfo is here :-)

**Comment by kilasuit** [comment]
Sorry but this is an absolutely atrocious Idea and IMO should not be merged.

Please stop putting bandaids around actual problems that will then lead to further issues down the line.

At the end of the day the AzureRM PowerShell team found this issue with their wise minds to change their module name in the first place, and is inherently their issue, and whilst this issue affects any module with 2 chars for the name this is not the correct long term fix to implement


@lzybkr first suggestion is a much better suggestion for the longer term and that should be the one that is favoured especially for those that may want an `IT` Module or an `EN` module in future.

**Comment by lzybkr** [comment]
In case it wasn't clear, my list was ordered by my preference.

**Comment by rjmholt** [comment]
Ok, I'll see if I can rework this to account for depth. The problem I imagine might occur there is with nested modules, particularly in `Get-Module -ListAvailable -All`.

The caching idea doesn't help the performance scenario this optimisation was introduced to help I don't think, so going that route would mean a fair amount of complexity to do something slightly better than just removing the optimisation.

The whitelist option at least means you can always fix the problem by adding to the whitelist. In this situation, everything is a bandaid, but only one of the solutions is an expandable bandaid. But I agree that it's far from ideal.

@kilasuit FWIW putting your criticism more constructively would be appreciated. Comments in these repos can get heated, and seeing them every day they can start to add up.

**Comment by iSazonov** [comment]
@rjmholt Why you think that culture name caching doesn't help?
Please look all culture name formats with `[cultureinfo]::GetCultures(0xf)`. There is names like `sr-Latn-RS`, `sms-FI`, `vo-001`, `zgh`. Seems a cache would be more simple and faster then a parser or regex.

**Comment by rjmholt** [comment]
The caching only comes *after* a module has determined to be a module. At first run, you have to read *every* directory and only add them to the cache if they are both (1) a module and (2) a culture name. 

The cache will slow down first startup and speed up later runs. But the whole culture name skip optimisation was made to improve startup time. So going the cache route will undo the original optimisation scenario.

There's nothing wrong with the idea though -- it's probably worth adding.

**Comment by iSazonov** [comment]
My suggestion is to filter directories here https://github.com/PowerShell/PowerShell/blob/4f8814031bf7c4c13bf04c376565a077ca9be2fd/src/System.Management.Automation/engine/Modules/ModuleUtils.cs#L90
Also we could use predefined culture name list (without call GetCultures() at start time because the list is changed rarely.
And we could use DirectoryInfo/FileInfo classes to exclude extra kernel calls if we want increase startup performance.

**Comment by adityapatwardhan** [comment]
@rjmholt Please resolve merge conflict.

**Comment by lzybkr** [comment]
@daxian-dbw - that could still require enumerating many files which isn't always fast. I think the depth check is perfectly reasonable.

**Comment by daxian-dbw** [comment]
> that could still require enumerating many files which isn't always fast.

@lzybkr Totally agree.
@rjmholt's concern on the depth check is when `-All` is specified. In that case, we want to find all modules nested in a module folder, where we will need to check the `az` folder under a module folder, and the depth check is not sufficient anymore. I admit that `-All` is not commonly used, but since we want to have a general fix, I guess we need to take that switch into account as well.

**Comment by lzybkr** [comment]
Does `-All` even work? I tried it in Windows PowerShell saw no results for as long as I was willing to wait.

**Comment by vexx32** [comment]
Yep. 

Takes about 1.5x as long as PowerShell 6.1, but Windows PowerShell does work. Both of them successfully listed all modules with `Get-Module -List -All`

**Comment by mklement0** [comment]
To perhaps offer a solution for future modules:

We could add support for placing localization folders in a dedicated subfolder of a module's root folder with a then-reserved name (the introduction of which can in itself be problematic), such as `Resources`.

Modules written for that new convention then wouldn't have the ambiguity problem, and the presence of the reserved folder would mean that everything else can assumed to be non-localization folders.

If there are concerns about folder-hierarchy depth / path length, the suboptimal alternative would be a reserved folder-name extension, the way that macOS uses `.lproj`, for instance (though the `.app` bundles there _also_ use a `Resources` subfolder).

**Comment by rjmholt** [comment]
@mklement0 +1 on dedicated folder (or a manifest configuration to avoid being opinionated, but that's a backcompat breaker I suppose). A similar implementation detail to this one [lead to dotfiles on UNIX](https://plus.google.com/+RobPikeTheHuman/posts/R58WgWwN9jp) and the official solution is to have a [standard directory/location for those files](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

**Comment by iSazonov** [comment]
I see that _CoreFX_ files is placed in locale folders. Currently it is Microsoft.CodeAnalysis*.

**Comment by adityapatwardhan** [comment]
@rjmholt Please have a look at the merge conflict.

**Comment by rjmholt** [comment]
> @rjmholt Robert Holt FTE Could you indicate the reason for closing?

We haven't settled on an implementation and this branch is stale.

It will be easier to open a new PR with a minimal change I think.

**Pull request closed without merging**